### ISSUE TITLE
Broken pattern span for identifier parameters in macros

### ISSUE THREAD

**Issue by panicbit** [rust-lang/rust#32072]
This kind of macro pattern seems to mess up the span:

``` rust
macro_rules! bad {
    ($s:ident whatever) => {
        {
            let $s = 0;
            *&mut $s = 0;
        }
    }
}

fn main() {
    bad!(foo whatever);
}
```

[Playpen](http://is.gd/ZtC5C0)

Compiling it results in a messed up error message:

```
<anon>:11:14: 11:22 error: cannot borrow immutable local variable `foo` as mutable
<anon>:11     bad!(foo whatever);
                       ^~~~~~~~
<anon>:11:5: 11:24 note: in this expansion of bad! (defined in <anon>)
<anon>:11:14: 11:22 help: to make the local variable mutable, use `mut` as shown:
<anon>:       bad!(foo mut whatever);
<anon>:11:5: 11:24 note: in this expansion of bad! (defined in <anon>)
error: aborting due to previous error
```

**Comment by Mark-Simulacrum** [comment]
This is almost certainly not what we want today.. Or at least, it looks really odd.

```
error[E0596]: cannot borrow immutable local variable `foo` as mutable
  --> test.rs:11:10
   |
5  | |             *&mut $s = 0;
   | |____________________^ cannot borrow mutably
...
11 |       bad!(foo whatever);
   |  _____-----^-------------
   | |     |
   | |     in this macro invocation

error: aborting due to previous error(s)
```

**Comment by estebank** [comment]
CC #23480.

**Comment by estebank** [comment]
The current output is beyond bad because of the new suggestion:

```
error[E0596]: cannot borrow immutable local variable `foo` as mutable
  --> src/main.rs:5:19
   |
4  |                let $s = 0;
   |   _________________-
5  |  |             *&mut $s = 0;
   |  |___________________^
6  | ||         }
7  | ||     }
8  | || }
9  | ||
10 | || fn main() {
11 | ||     bad!(foo whatever);
   | ||     -------^-----------
   | ||_____|______|
   | |______|______consider changing this to `mut $s = 0;
            *&mut $s = 0;
        }
    }
}

fn main() {
    bad!(foo`
   |        |      cannot borrow mutably
   |        in this macro invocation
```

Fixed in #47942.
